File: Spring 2020/Pig_Latin/Python/main.py
import re

def pig_latin(sentence):
    sentence = sentence.split(' ')
    result = ''
    for word in sentence:
        first_vowel = re.findall(r'[aeiouAEIOU]', word)[0]
        index = word.find(first_vowel)
        result += word + 'way ' if not index else word[index:] + word[:index] + 'ay '
    return result[:len(result) - 1]

def pig_latin_brute_force(sentence):
    result = ''
    vowel = {'a' : True, 'e' : True, 'i' : True, 'o' : True, 'u' : True}
    sentence = sentence.split(' ')
    for word in sentence:
        if word[0].lower() in vowel:
            result += word + 'way '
        else:
            i = 1
            while i < len(word) and not(word[i] in vowel):
                i += 1
            result += word[i:] + word[:i] + 'ay ' 

    return result[:len(result) - 1]

print(pig_latin('   Hello   There YOu    '))

File: Spring 2020/Pig_Latin/Java/Main.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.*;

class Main {
    public static void main(String[] args) {
        System.out.println(pigLatin("Everyone thinks that smiling will solve the problems of the world"));
        System.out.println(pigLatinBruteForce("Everyone thinks that smiling will solve the problems of the world"));
    }
    
    public static String pigLatin(String sentence) {
        Matcher matcher;
        int index;

        String[] words = sentence.split(" ");
        String result = "";
        Pattern pattern = Pattern.compile("[aeiouAEIOU]");

        for (String word: words) {
            matcher = pattern.matcher(word);
            if (matcher.find()) {
                result += ((index = word.indexOf(matcher.group(0))) != 0) ? word.substring(index) + word.substring(0, index) + "ay " : word + "way "; 
            }
        }
        return result.substring(0, result.length() - 1);
    }

    public static String pigLatinBruteForce(String sentence) {
        String result = "";
        String[] words = sentence.split(" ");
        Hashtable<String, Boolean> vowels = new Hashtable<String, Boolean>();

        int j;

        vowels.put("a", true);
        vowels.put("e", true);
        vowels.put("i", true);
        vowels.put("o", true);
        vowels.put("u", true);
        vowels.put("A", true);
        vowels.put("E", true);
        vowels.put("I", true);
        vowels.put("O", true);
        vowels.put("U", true);

        for (int i = 0; i < words.length; i++) {
            if (vowels.containsKey(String.valueOf(words[i].charAt(0)))) {
                result += words[i] + "way ";
            } else {
                for (j = 0; j < words[i].length() && !vowels.containsKey(String.valueOf(words[i].charAt(j))); j++);
                result += words[i].substring(j) + words[i].substring(0, j) + "ay ";
            }
        }
        return result.substring(0, result.length() - 1);
    }
}
